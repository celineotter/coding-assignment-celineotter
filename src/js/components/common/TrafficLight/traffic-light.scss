.traffic-light-container {
  display: flex;
  justify-content: center;
}

.circle {
  width: 100px;
  height: 100px;
  border-radius: 50px;
  background-color:black;
  position: absolute;
}

.circle0 {
    top:1%;
}

.circle1 {
    top:34%;
}

.circle2 {
    top:67%;
}

.red {
  background-color: red;
}

.yellow {
  background-color: yellow;
}

.green {
  background-color: green;
}

.blue {
  background-color: blue;
}

.orange {
  background-color: orange;
}

.purple {
  background-color: purple;
}

.square {
  width: 105px;
  height: 325px;
  background-color: white;
  position: absolute;
  border-style: solid;
  border-color: black;
  background-color: grey;
  z-index: 1;
}