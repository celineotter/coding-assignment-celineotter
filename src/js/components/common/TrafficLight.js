import React, { Component } from 'react';

import './traffic-light.css';

const DEFAULT_COLORS = [ 'red', 'yellow', 'green' ];
let cancelTimeOut;

class TrafficLight extends Component {
  constructor(props) {
    super(props);
    this.state = {
      colors: this.props.colors ? this.props.colors : DEFAULT_COLORS,
      colorIndex: 0,
    };
  }

  componentWillUnmount() {
    cancelTimeOut();
  }

  componentDidMount() {
    this.handleLightChange();
  }

  handleLightChange = () => {
    const tempIndex = this.state.colorIndex + 1;
    const newIndex = tempIndex > this.state.colors.length - 1 ? 0 : tempIndex;
    cancelTimeOut = setTimeout(() => {
      this.setState({ colorIndex: newIndex }, this.handleLightChange);
    }, 1000);
  }

  render() {
    return (
      <div>
        <div styleName="traffic-light-container">
          <div styleName="square">
            {this.state.colors.map((color, index) => {
              const activeLight = index === this.state.colorIndex;
              const colorStyle = activeLight ? color : '';
              const style = `circle circle${index} ${colorStyle}`;

              return (
                <div styleName={style} key={`light${index}`} />
              );
            })}
          </div>
        </div>
      </div>

    );
  }
}

export default TrafficLight;
