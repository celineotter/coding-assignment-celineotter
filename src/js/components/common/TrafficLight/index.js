// @flow

import React, { Component } from 'react';

import './traffic-light.css';

type colorsType = Array<string>;

const DEFAULT_COLORS: colorsType = [ 'red', 'yellow', 'green' ];

let cancelTimeOutId;

type StateProps = {
  colors: ?colorsType,
};

type TrafficLightState = {
  colorIndex: ?number,
};

class TrafficLight extends Component<StateProps, TrafficLightState> {
  constructor(props: StateProps) {
    super(props);

    this.state = {
      colorIndex: 0,
    };
  }

  componentWillUnmount() {
    this.clearTimers();
  }

  componentDidMount() {
    this.handleLightChange();
  }

  clearTimers() {
    clearTimeout(cancelTimeOutId);
  }

  getColors() {
    return this.props.colors || DEFAULT_COLORS;
  }

  handleLightChange = () => {
    const tempIndex = this.state.colorIndex + 1;
    const newIndex = tempIndex > this.getColors().length - 1 ? 0 : tempIndex;

    cancelTimeOutId = setTimeout(() => {
      this.setState({ colorIndex: newIndex }, this.handleLightChange);
    }, 1000);
  }

  render() {
    return (
      <div>
        <div styleName="traffic-light-container">
          <div styleName="square">
            {this.getColors().map((color, index) => {
              const activeLight = index === this.state.colorIndex;
              const style = [ 'circle', `circle${index}`, activeLight ? color : '' ].join(' ');

              return (
                <div testId="light" styleName={style} key={`light${index}`} />
              );
            })}
          </div>
        </div>
      </div>
    );
  }
}

export default TrafficLight;
