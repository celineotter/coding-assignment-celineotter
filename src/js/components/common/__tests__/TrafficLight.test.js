import React from 'react';
import { shallow } from 'enzyme';

import TrafficLight from '@common/TrafficLight';

describe('TrafficLight', () => {
  let wrapper;

  const getWrapper = (options) => {
    return shallow(<TrafficLight {...options} />);
  };

  const checkLightIsActive = (lights, lightIndex, color) => {
    return shallow(lights.get(lightIndex)).props().className.includes(color);
  };

  beforeEach(() => {
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.clearAllTimers();
  });

  it('renders the component', () => {
    wrapper = getWrapper();
    expect(wrapper).toMatchSnapshot();
  });

  it('clears the timers on unmount', () => {
    clearTimeout = jest.fn(); // eslint-disable-line
    wrapper = getWrapper();
    wrapper.instance().componentWillUnmount();
    expect(clearTimeout).toHaveBeenCalledTimes(1);
  });

  describe('first render', () => {
    it('displays only the red color as active', () => {
      wrapper = getWrapper();
      const lights = wrapper.find({ testId: 'light' });

      expect(checkLightIsActive(lights, 0, 'red')).toBe(true);
      expect(checkLightIsActive(lights, 1, 'yellow')).toBe(false);
      expect(checkLightIsActive(lights, 2, 'green')).toBe(false);
    });
  });

  describe('following 1 second', () => {
    it('displays only the yellow color as active', () => {
      wrapper = getWrapper();
      jest.advanceTimersByTime(1000);

      wrapper.update();
      const lights = wrapper.find({ testId: 'light' });

      expect(checkLightIsActive(lights, 0, 'red')).toBe(false);
      expect(checkLightIsActive(lights, 1, 'yellow')).toBe(true);
      expect(checkLightIsActive(lights, 2, 'green')).toBe(false);
    });
  });

  describe('following 2 second', () => {
    it('displays only the yellow color as active', () => {
      wrapper = getWrapper();
      jest.advanceTimersByTime(2000);

      wrapper.update();
      const lights = wrapper.find({ testId: 'light' });

      expect(checkLightIsActive(lights, 0, 'red')).toBe(false);
      expect(checkLightIsActive(lights, 1, 'yellow')).toBe(false);
      expect(checkLightIsActive(lights, 2, 'green')).toBe(true);
    });
  });

  describe('following 3 second', () => {
    it('starts the cycle over and displays only the red color as active', () => {
      wrapper = getWrapper();
      jest.advanceTimersByTime(3000);

      wrapper.update();
      const lights = wrapper.find({ testId: 'light' });

      expect(checkLightIsActive(lights, 0, 'red')).toBe(true);
      expect(checkLightIsActive(lights, 1, 'yellow')).toBe(false);
      expect(checkLightIsActive(lights, 2, 'green')).toBe(false);
    });
  });
});
