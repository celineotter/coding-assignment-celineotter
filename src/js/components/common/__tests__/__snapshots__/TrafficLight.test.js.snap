// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TrafficLight renders the component 1`] = `
<div>
  <div
    className="src-js-components-common-___traffic-light__traffic-light-container___1gFeD"
  >
    <div
      className="src-js-components-common-___traffic-light__square___c6kT8"
    >
      <div
        className="src-js-components-common-___traffic-light__circle___37Rk0 src-js-components-common-___traffic-light__circle0___2pVep src-js-components-common-___traffic-light__red___39QGm"
        key="light0"
        testId="light"
      />
      <div
        className="src-js-components-common-___traffic-light__circle___37Rk0 src-js-components-common-___traffic-light__circle1___2dTvV"
        key="light1"
        testId="light"
      />
      <div
        className="src-js-components-common-___traffic-light__circle___37Rk0 src-js-components-common-___traffic-light__circle2___2H0Ra"
        key="light2"
        testId="light"
      />
    </div>
  </div>
</div>
`;
